connectivity/tests/common.go

```diff
@@ -48,3 +54,81 @@ func hasAllLabels(labelsContainer labelsContainer, filters map[string]string) bo
 	}
 	return true
 }
+
+type retryCondition struct {
+	podLabels map[string]string
+	all       bool
+	destPort  uint32
+	destIP    string
+}
+
+// CurlOptions returns curl retry option or empty slice depending on retry conditions
+// if all value set it means always return retry options
+// for pod-to-cidr all or destIP is checked
+// for pod-to-world all, destPort and destPort + pod label is checked
+func (rc *retryCondition) CurlOptions(peer check.TestPeer, ipFam check.IPFamily, pod check.Pod, params check.Parameters) []string {
+	opts := []string{}
+
+	if connectRetry := params.Retry; connectRetry > 0 {
+		opts = append(opts, "--retry", strconv.FormatInt(int64(connectRetry), 10),
+			"--retry-all-errors") // add --retry-all-errors to retry on all possible errors
+
+		// only add --retry-delay if --retry is enabled
+		if retryDelay := params.RetryDelay.Seconds(); retryDelay > 0.0 {
+			opts = append(opts, "--retry-delay", strconv.FormatFloat(retryDelay, 'f', -1, 64))
+		}
+	}
+
+	// no retry options given return early
+	if len(opts) == 0 {
+		return opts
+	}
+	// all conditions is set return early
+	if rc.all {
+		return opts
+	}
+	// destIP is checked with pod-to-cidr cases
+	if peer.Address(ipFam) == rc.destIP {
+		return opts
+	}
+	// destPort and podLabel are checked with pod-to-world case
+	if peer.Port() == rc.destPort {
+		for n, v := range rc.podLabels {
+			if !pod.HasLabel(n, v) {
+				return []string{}
+			}
+		}
+		return opts
+	}
```

Comment:
It feels a bit strange this this is essentially `all OR destIP matches OR (desPort matches AND label matches)`. Though it currently doesn't have a practical consequence, I'd say it's more straight forward to apply the options if either `all` is true OR all specified options are true, i.e. the `AND` of all specified options. Something like:

```go
func (rc *retryCondition) CurlOptions(peer check.TestPeer, ipFam check.IPFamily, pod check.Pod, params check.Parameters) []string {                                                                                                                
        if params.Retry == 0 { 
                return []string{}
        } 

        opts := []string{
                "--retry", strconv.FormatInt(int64(connectRetry), 10),
                "--retry-all-errors", // add --retry-all-errors to retry on all possible errors
        } 

        if retryDelay := params.RetryDelay.Seconds(); retryDelay > 0.0 { 
                opts = append(opts, "--retry-delay", strconv.FormatFloat(retryDelay, 'f', -1, 64))
        } 
        if rc.all { 
                return opts
        } 
        if rc.destIP != "" && peer.Address(ipFam) != rc.destIP { 
                return []string{}
        } 
        if rc.destPort != 0 && peer.Port() != rc.destPort { 
                return []string{}
        } 
        for n, v := range rc.podLabels { 
                if !pod.HasLabel(n, v) { 
                        return []string{}
                } 
        } 

        return opts
}
```

That way it's independent of the test cases these options are currently used with and allows for easier extensibility in the future.

What do you think?

Reply:
This implementation is much more elegant 👍 but it is missing one thing if no options are set it still sets the retry options. So I added a check earlier to return in this case
```
	if !rc.all && rc.destIP == "" && rc.destPort == 0 {
		return []string{}
	}
```